### PR TITLE
when storing references, check that they are compatible with our garb…

### PR DESCRIPTION
**Pull request by galganif**
…age collection impl

### CODE REVIEW

**Comment by galganif** [comment]
needs more work for reference mode keys

**Comment by galganif** [comment]
Ready for review now.

File: java/arcs/core/entity/StorageAdapter.kt [hunk 1]
```diff
@@ -27,6 +30,29 @@ sealed class StorageAdapter<T : Storable, R : Referencable> {
 
     /** Checks if the [Storable] is expired (its expiration time is in the past). */
     abstract fun isExpired(value: T): Boolean
+
+    @Throws(IllegalStateException::class)
```
Comment (jasonwyatt) [inline java/arcs/core/entity/StorageAdapter.kt:34]: This isn't really necessary unless we want java callers to have to catch the exception.

File: java/arcs/core/entity/StorageAdapter.kt [hunk 2]
```diff
@@ -27,6 +30,29 @@ sealed class StorageAdapter<T : Storable, R : Referencable> {
...
+            entitiesKey = handleKey.backingKey
+        }
+        // If we are pointing to an entity in the database, we should also be using a storage key in
+        // the same database. Otherwise the entity may be garbage collected from the database and
+        // the reference become invalid.
+        if (referencedKey is DatabaseStorageKey) {
+            if (entitiesKey !is DatabaseStorageKey || entitiesKey.dbName != referencedKey.dbName)
+                throw IllegalStateException("References to database entity should only be stored" +
+                    " in the same database. You are using $handleKey to store a reference that" +
+                    " lives in the ${referencedKey.dbName} database.")
```
Comment (jasonwyatt) [inline java/arcs/core/entity/StorageAdapter.kt:53]:
```suggestion
            check(entitiesKey is DatabaseStorageKey && entitiesKey.dbName == referencedKey.dbName) {
                "References to database entity should only be stored in the same database. " + 
                    "You are using $handleKey to store a reference that lives in the " + 
                    "${referencedKey.dbName} database."
            }
```

`check` throws an IllegalStateException.

**Pull request merged**